File: Estudos/3 prova/beecrownd/Percurso em Arvore por Nivel 1466/main.c
#include <stdio.h>
#include <stdlib.h>

typedef struct No {
    int valor;
    struct No *esq;
    struct No *dir;
} No;

No* novoNo(int x) {
    No *n = (No*) malloc(sizeof(No));
    n->valor = x;
    n->esq = NULL;
    n->dir = NULL;
    return n;
}

No* inserir(No *raiz, int x) {
    if (raiz == NULL) {
        raiz = novoNo(x);
    } else if (x < raiz->valor) {
        raiz->esq = inserir(raiz->esq, x);
    } else if (x > raiz->valor) {
        raiz->dir = inserir(raiz->dir, x);
    }
    return raiz;
}

void liberar(No *raiz) {
    if (raiz != NULL) {
        liberar(raiz->esq);
        liberar(raiz->dir);
        free(raiz);
    }
}

void percursoNivel(No *raiz, int n) {
    No *fila[600];
    int inicio = 0;
    int fim = 0;
    int primeiro = 1;

    if (raiz != NULL) {
        fila[fim++] = raiz;
    }

    while (inicio < fim) {
        No *atual = fila[inicio++];
        if (!primeiro) {
            printf(" ");
        }
        printf("%d", atual->valor);
        primeiro = 0;

        if (atual->esq != NULL) {
            fila[fim++] = atual->esq;
        }
        if (atual->dir != NULL) {
            fila[fim++] = atual->dir;
        }
    }

    printf("\n");
}

int main() {
    int casos;
    int i, j;

    if (scanf("%d", &casos) != 1) {
        return 0;
    }

    for (i = 1; i <= casos; i++) {
        int n;
        scanf("%d", &n);

        No *raiz = NULL;

        for (j = 0; j < n; j++) {
            int x;
            scanf("%d", &x);
            raiz = inserir(raiz, x);
        }

        printf("Case %d:\n", i);
        percursoNivel(raiz, n);
        printf("\n");

        liberar(raiz);
    }

    return 0;
}

File: Estudos/3 prova/beecrownd/Percurso em Arvore por Nivel 1466/Main.java
import java.util.Scanner;

class No {
    int valor;
    No esq, dir;

    public No(int v) {
        this.valor = v;
        this.esq = null;
        this.dir = null;
    }
}

class Arvore {
    No raiz;

    public Arvore() {
        raiz = null;
    }

    public void inserir(int x) {
        raiz = inserirRec(raiz, x);
    }

    private No inserirRec(No i, int x) {
        if (i == null) {
            i = new No(x);
        } else if (x < i.valor) {
            i.esq = inserirRec(i.esq, x);
        } else if (x > i.valor) {
            i.dir = inserirRec(i.dir, x);
        }
        return i;
    }

    public void percursoNivel() {
        if (raiz == null) {
            System.out.println();
            return;
        }

        No[] fila = new No[600]; 
        int inicio = 0;
        int fim = 0;

        fila[fim++] = raiz;

        boolean primeiro = true;

        while (inicio < fim) {
            No atual = fila[inicio++];

            if (!primeiro) {
                System.out.print(" ");
            }
            System.out.print(atual.valor);
            primeiro = false;

            if (atual.esq != null) {
                fila[fim++] = atual.esq;
            }
            if (atual.dir != null) {
                fila[fim++] = atual.dir;
            }
        }

        System.out.println();
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner ent = new Scanner(System.in);

        int casos = ent.nextInt();

        for (int c = 1; c <= casos; c++) {
            int n = ent.nextInt();

            Arvore arvore = new Arvore();

            for (int i = 0; i < n; i++) {
                int x = ent.nextInt();
                arvore.inserir(x);
            }

            System.out.println("Case " + c + ":");
            arvore.percursoNivel();
            System.out.println();
        }

        ent.close();
    }
}
